controllers/iamauthpolicy_controller.go

```diff
@@ -2,67 +2,80 @@ package controllers
+	model "github.com/aws/aws-application-networking-k8s/pkg/model/lattice"
 	"github.com/aws/aws-application-networking-k8s/pkg/utils/gwlog"
 
-	k8serr "k8s.io/apimachinery/pkg/api/errors"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
 )
 
 type IAMAuthPolicyController struct {
-	log    gwlog.Logger
-	client client.Client
+	log       gwlog.Logger
+	client    client.Client
+	policyMgr deploy.IAMAuthPolicyManager
 }
 
-func RegisterIAMAuthPolicyController(log gwlog.Logger, mgr ctrl.Manager) error {
+func RegisterIAMAuthPolicyController(log gwlog.Logger, mgr ctrl.Manager, cloud pkg_aws.Cloud) error {
 	controller := &IAMAuthPolicyController{
 		log:    log,
 		client: mgr.GetClient(),
+		policyMgr: deploy.IAMAuthPolicyManager{
+			Cloud: cloud,
+		},
 	}
 	err := ctrl.NewControllerManagedBy(mgr).
 		For(&anv1alpha1.IAMAuthPolicy{}).
+		WithEventFilter(predicate.GenerationChangedPredicate{}).
 		Complete(controller)
 	return err
 }
 
 func (c *IAMAuthPolicyController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	policySpec := &anv1alpha1.IAMAuthPolicy{}
+	err := c.client.Get(ctx, req.NamespacedName, policySpec)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
 	c.log.Infow("reconcile", "req", req)
 
-	policy := &anv1alpha1.IAMAuthPolicy{}
-	err := c.client.Get(ctx, req.NamespacedName, policy)
-	if !k8serr.IsNotFound(err) {
-		return ctrl.Result{}, err
+	policyModel, err := model.NewIAMAuthPolicy(policySpec)
+	if err != nil {
+		return ctrl.Result{RequeueAfter: time.Hour}, err // good candidate for terminal error
 	}
 
-	switch policy.Spec.TargetRef.Kind {
-	case "Gateway":
-		err = c.reconcileGateway(ctx, policy)
-		break
-	case "HTTPRoute":
-	case "GRPCRoute":
-		err = c.reconcileRoute(ctx, policy)
-		break
-	default:
-		err = fmt.Errorf("unsupported targetRef type, req=%s, kind=%s",
-			req, policy.Spec.TargetRef.Kind)
+	authPolicyFinalizer := "iamauthpolicy.k8s.aws/resources"
+
+	if policySpec.DeletionTimestamp.IsZero() {
+		policyStatus, err := c.policyMgr.Put(ctx, policyModel)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		policySpec.Annotations["ARN"] = policyStatus.Resource.ARN.String()
+		if !controllerutil.ContainsFinalizer(policySpec, authPolicyFinalizer) {
+			controllerutil.AddFinalizer(policySpec, authPolicyFinalizer)
+		}
+		c.log.Infof("updated policy for name=%s, latticeId=%s", policyModel.ResourceName, policyStatus.Resource.Id)
+	} else {
+		if controllerutil.ContainsFinalizer(policySpec, authPolicyFinalizer) {
+			err = c.policyMgr.Delete(ctx, policyModel)
+			if err != nil {
+				return ctrl.Result{}, pkg_aws.IgnoreNotFound(err)
+			}
+			controllerutil.RemoveFinalizer(policySpec, authPolicyFinalizer)
+			c.log.Infof("deleted policy for name=%s", policyModel.ResourceName)
```

Comment:
Should we not be logging the policy namespaced name? Or is that the resourceName?

Reply:
it's LatticeServiceName, name+namespace truncated
https://github.com/aws/aws-application-networking-k8s/blob/main/pkg/utils/common.go#L38